src/driver/linux_resource/syscall_aarch64.c

```diff
@@ -57,70 +233,211 @@ static typeof(aarch64_insn_adrp_get_offset) *aarch64_insn_adrp_get_offset_sym;
  * be sure it will never break, so there are as many safety checks here as possible.
  */
 
+/* Modern kernels way (since 5.4) (tested on 5.15.83):
+ * - Instead of `el_sync` we look for `el0t_64_sync`. Some new code was added
+ *   to the exception handler. See find_el_sync() for more comments.
+ * - `el0t_64_sync` jumps to C function el0t_64_sync_handler(), which calls
+ *   specific function based on the value of esr_el1 system register. We look
+ *   for `el0_svc` here. See find_el_svc_entry() for more comments.
+ * - `el0_svc()` calls `do_el0_svc()`, which calls `el0_svc_common()` and passes
+ *   pointer to the syscall table as an argument. See find_syscall_table_via_vbar()
+ *   for more comments.
+ */
+
 static ci_uint8 *
 find_el_sync(void)
 {
   ci_uint8 *vectors;
   ci_uint32 insn;
-  ci_uint8 *el_sync = efrm_find_ksym("el0_sync");
+  ci_uint8 *el_sync = NULL;
 
+#ifdef EFRM_HAVE_NEW_KALLSYMS
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 13, 0)
+  el_sync = efrm_find_ksym("el0t_64_sync");
+#else
+  el_sync = efrm_find_ksym("el0_sync");
+#endif
+#endif
   if (el_sync != NULL)
     return el_sync;
 
   vectors = (ci_uint8 *)read_sysreg(vbar_el1);
   vectors += 8 * 128;
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 28)
+  /* Skip code that was added in 5.15.28:
+   * 14000003        b       40c <vectors+0x40c>
+   */
+  vectors += AARCH64_INSN_SIZE;
+#endif
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 15, 0)
+  /* Skip code that was added in 4.15:
+   * d53bd07e        mrs     x30, tpidrro_el0
+   * d51bd07f        msr     tpidrro_el0, xzr
+   */
+  vectors += 2 * AARCH64_INSN_SIZE;
+#endif
+
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 14, 0)
   vectors += sizeof(ci_uint32);
 #if defined(CONFIG_VMAP_STACK)
   vectors += 5 * sizeof(ci_uint32);
 #endif
 #endif
-  if (aarch64_insn_read_sym(vectors, &insn)) {
+  if (ci_aarch64_insn_read(vectors, &insn)) {
     EFRM_WARN("%s: cannot read vbar_el1", __func__);
     return NULL;
   }
   if (!aarch64_insn_is_b(insn)) {
     EFRM_WARN("%s: el0_sync entry is not a branch", __func__);
     return NULL;
   }
-  el_sync = vectors + aarch64_get_branch_offset_sym(insn);
+  el_sync = vectors + ci_aarch64_get_branch_offset(insn);
+
+  TRAMP_DEBUG("%s:%d: 'b el_sync' = %px (%x)", __FUNCTION__, __LINE__, vectors,
+              insn);
+  TRAMP_DEBUG("%s:%d: el_sync = %px", __FUNCTION__, __LINE__, el_sync);
 
   return el_sync;
 }
 
+static ci_uint8 *
+find_next_bl(ci_uint8 *ptr)
+{
+  ci_uint32 insn;
+
+  while (1) {
+    CI_AARCH64_INSN_READ(ptr, insn);
+    if (aarch64_insn_is_bl(insn))
```

Comment:
nit: (This is a matter of preference, really) Shall we put it inside `while` like `while (!aarch64_insn_is_bl(insn))` here and other occurrences?

Reply:
Fixed